### ISSUE TITLE
common utils.R 

### ISSUE THREAD

**Issue by MOshima-PIFSC** [r4ss/r4ss#617]
In yesterday's discussion with the r4ss and ss3diags teams, Ian mentioned possibly creating a separate utils package for both packages to use. I like the idea but a separate, third package might be difficult to maintain. What about a common utils.R script where we keep basic functions? And then, if possible, we could set up a Github action so that if changes were pushed to the file from one repo it would be pushed to the other as well. I'm not sure exactly how to set that up but maybe [something like this](https://github.com/marketplace/actions/push-a-file-to-another-repository). 

Functions that are already in the ss3diags utils.R are:
*  sspar (used to make multiplots) 
* SSdiagsTime2Year (to convert non-annual into annual time steps)
* rc (same as r4ss::rich.colors.short)
* legendfun

Some candidate functions from r4ss might be:
* rich.colors.short (we can change rc function in ss3diags to rich.colors.short)
* matchfun
* matchfun2
* pngfun (ss3diags uses this function too but it is not in the utils yet)
* get_dat_new_name 

This was just an idea I had. Would be interested to hear others thoughts on this!
@efletcherPIFSC @fecor21 @Henning-Winker

**Comment by k-doering-NOAA** [comment]
What do you think about just putting the functions in r4ss, @MOshima-PIFSC? I think this makes sense given that ss3diags already depends on r4ss. Changes to accomodate ss3diags could just be pushed directly to r4ss.

**Comment by k-doering-NOAA** [comment]
Meant to say, I also like the idea of reducing the duplicated code!

**Comment by MOshima-PIFSC** [comment]
@k-doering-NOAA I'm totally fine with that idea too!

**Comment by k-doering-NOAA** [comment]
@MOshima-PIFSC great! Would you (and/or Eric) feel comfortable forking r4ss and submitting a PR for the changes you would like? That way @iantaylor-NOAA can take a look and make sure it doesn't break anything in r4ss.

Otherwise, I'm happy to coordinate with you and figure out how to best accomplish this.

**Comment by roliveros-ramos** [comment]
My two cents:
Create utils-r4ss.R and utils-ss3diags.R, so one file is always a pasive copy in the other repo. While the automatic merge of git is quite good, non overlaping files can save some problems later.

**Comment by iantaylor-NOAA** [comment]
Thanks for all the ideas. 
Given that ss3diags is already dependent on r4ss, I do like the simplicity of having all these functions in just one place rather than dealing with copies. If that seems necessary in the future, we could follow the suggestion that @roliveros-ramos just made.

Therefore, just putting these into r4ss seems easiest for now. I'm happy to have PIFSC folks take the lead on pulling functions into a separate file within r4ss. 

Here are a few suggestions to consider:
* @k-doering-NOAA already created a `helper_fxns.R` file so we could just use that one to store these things (and potentially rename to `utils.R` if people like that better: https://github.com/r4ss/r4ss/blob/main/R/helper_fxns.R.
* I'd be happy to work on pulling out `pngfun()` myself. I started that last year in the `section508` branch but wasn't able to finish before getting wrapped up in assessment work (see changes here: https://github.com/r4ss/r4ss/compare/main...section508). I'd be happy to finish that task which would probably be faster than starting fresh or somebody else trying to complete my unfinished work.
* At some point it may make sense to have all users install both r4ss and ss3diags, but at this stage where ss3diags hasn't been used as widely it's probably fine to let r4ss users choose to install it only if they want to try it out.

**Comment by fecor21** [comment]
Sounds like a good plan to me.

**Comment by MOshima-PIFSC** [comment]
@iantaylor-NOAA that sounds good to me. @efletcherPIFSC and I can work on adding the functions to the `helper_fxns.R`

**Comment by efletcherPIFSC** [comment]
Hello all,

'helper_fxns.R' functions with the -fun suffix is broad and subjectively contradict with the code documentation.  The following are the proposed renamed functions:   

legendfun => add_legend     (Adds legend to plots)
matchfun => match_report_line    (Extract line number from Report.sso by keyword)
matchfun2 => match_report_table  (Extract table from Report.sso by keyword)
pngfun => save_ssplot_png    (Save as PNG)

**Comment by iantaylor-NOAA** [comment]
Thanks @efletcherPIFSC, these are all good ideas. Funny how terrible function names can stick around forever until somebody suggests something better.

I have two questions for you about pngfun: 
1. how about just `save_png()` instead of `save_ssplot_png()` as the new function name?
2. when it's embedded within a plotting function, this helper function can use local variables within that function, but pulling it out requires specifying all the arguments explicitly. That is, before it could be called as `pngfun(file = "somefile.png")` but now it needs to be 
`      plotinfo <- pngfun(
        plotinfo = plotinfo, file = file, plotdir = plotdir, pwidth = pwidth,
        pheight = pheight, punits = punits, res = res, ptsize = ptsize,
        caption = caption
      )
`
I'm OK with that lengthier list of inputs, but is that the best practices in cases where it is called many times within a function (e.g. 24 times in https://github.com/r4ss/r4ss/blob/main/R/SSplotBiology.R)?

**Comment by efletcherPIFSC** [comment]
Thanks @iantaylor-NOAA.

1. Renaming pngfun to `save_png()` is more natural than `save_ssplots_png()`. My concern was name conflicts. Otherwise, I am on board with renaming it save_png.

2. In this case. One way, is I done is to wrap the graphical parameters in a list and pass it to the png function. I am unsure it is the most efficient way, however it works for me. Passing graphical parameters through the ellipsis (`...`) argument would simplify the code, however it may need validation.

**Comment by k-doering-NOAA** [comment]
I agree with the suggestions @efletcherPIFSC made for 2. It definitely is not a good practice to use variables that aren't explicitly passed to a function in R.

 Even though it is more verbose, I like specifying all the variables (or passing them in using a list).

**Comment by kellijohnson-NOAA** [comment]
@iantaylor-NOAA you can use `do.call()` and supply all the variables as a list that you bundle one time if you want. I also think that this is a good place to use the elipse functionality within R.

**Comment by iantaylor-NOAA** [comment]
Thanks for the feedback @efletcherPIFSC, @k-doering-NOAA and @kellijohnson-NOAA.
I like the idea of passing a list or moving to the use of ellipses, but I realized that I had already developed the current code with all the arguments explicitly passed to the function, so will leave any additional refinements to another day and another issue.

Here's a checklist adapted from the comments by @MOshima-PIFSC and @efletcherPIFSC above: https://github.com/r4ss/r4ss/issues/617#issue-1138823470 https://github.com/r4ss/r4ss/issues/617#issuecomment-1042330814.

- [ ] move some or all functions the ss3diags utils.R to r4ss helper_fxns.R, these include
  - [ ] sspar (used to make multiplots)
  - [ ] SSdiagsTime2Year (to convert non-annual into annual time steps)
  - [ ] rc (same as r4ss::rich.colors.short)
  - [ ] legendfun
- [ ] move functions from throughout r4ss into helper_fxns.R or use versions copied from ss3diags
  - [ ] rich.colors.short (we can change rc function in ss3diags to rich.colors.short)
  - [ ] legendfun => add_legend (Adds legend to plots)
  - [ ] matchfun => match_report_line (Extract line number from Report.sso by keyword)
  - [ ] matchfun2 => match_report_table (Extract table from Report.sso by keyword)
  - [x] pngfun => save_png (ss3diags uses this function too but it is not in the utils yet) [completed in #620]
- [ ] increment r4ss version number and add additional authors to DESCRIPTION and r4ss-package.R

@efletcherPIFSC and @MOshima-PIFSC, please comment if you want help on any of this

**Comment by k-doering-NOAA** [comment]
@iantaylor-NOAA @MOshima-PIFSC @efletcherPIFSC , wow, lots of work on this issue, great job! Is it complete or are there still lingering tasks? If it is complete, please feel free to close!

**Comment by MOshima-PIFSC** [comment]
Thanks @k-doering-NOAA and @iantaylor-NOAA, I forgot this issue was still open. I'm satisfied with everything for the time, I will let @efletcherPIFSC close the issue.

**Comment by efletcherPIFSC** [comment]
@k-doering-NOAA @iantaylor-NOAA I am satisfied with this: The actions with the solution has been addressed. 
However, I have to ask you to close this issue.

**Comment by k-doering-NOAA** [comment]
Thanks!
